### PR TITLE
feat(progress-indicator): add fitText prop and functionality to progress step

### LINKED ISSUE #5381: Progress Indicator Progress Step always truncates step label
### Summary

We are trying to implement a Progress Indicator but find the truncated text on labels to limit usability. As far as we could see there is no way to remove the truncation and have the steps fit the label size. 

This will be a design and development choice I believe. 

### Justification

Required by IBM designers, please contact me within IBM for more details.

### Desired UX and success metrics

The progress steps should display all the text input to the label if required.

### "Must have" functionality

The progress steps should display all the text input to the label if required.

### Specific timeline issues / requests

Yes, please contact me within IBM for more details. 

_NB: The Carbon team will try to work with your timeline, but it's not
guaranteed. The earlier you make a request in advance of a desired delivery
date, the better!_

### Available extra resources

I have created a pull request with the required changes: 

https://github.com/carbon-design-system/carbon/pull/5380

_Carbon is a collaborative system. We encourage teams to build components and
submit them for integration as either add-ons or core components._

### PR DESCRIPTION
**Pull request by samgoulden28**
Closes https://github.com/carbon-design-system/carbon/issues/5381

Allow the text label of a progress step to be non-truncated and the step element to grow to fit the typed text.

#### Changelog

**New**

Added fitText prop to ProgressStep component. 
Added a new Story to show the new fitText function.
Added new <CLASS>--no-max-width classes

**Changed**

Changed progress-step, progress-label and progress-line elements to optionally include the new --no-max-width class

**Removed**

Nothing

#### Testing / Reviewing

Running the storybook using `yarn storybook` and clicking Progress Indicator -> Fits Text will show the new functionality in action. You can then view the Storybook code to see how to add the new prop.

### CODE REVIEW

**Comment by tw15egan** [comment]
Could you create an issue to illustrate the problem this PR is trying to solve?

**Comment by samgoulden28** [comment]
@tw15egan Hi TJ, sure. I have created https://github.com/carbon-design-system/carbon/issues/5381.

As it says within the issue, please contact me within IBM if you require more details.

**Review by laurenmrice — CHANGES_REQUESTED** [review]
Each section in the progress indicator needs to be the same length. This is dependent on the length of the section with the longest word in this case would be 217px.

**Comment by samgoulden28** [comment]
Made the changes requested @laurenmrice using flex-grow to make all the steps the same size

**Comment by samgoulden28** [comment]
Additionally moved the prop from the ProgressStep to the ProgressIndicator in the storybook to make it consistent for all children without needing to specify fitText on all components

**Review by laurenmrice — CHANGES_REQUESTED** [review]
There should be 16px after the longest line of text and that is where the above rule should end for that section. Right now its too long, but they are all the same length which is good!

Ex:
Artboard

-------------------------------------------------------------------------------------------------
Now that I have access to the theme switcher, I did notice in the dark themes some colors are wrong. The blue rule lines need to be calling `interactive-04` . We can tackle that in another pr separately if that is easier.

File: packages/react/scripts/build.js [hunk 1]
```diff
@@ -28,7 +28,7 @@ try {
   exec(`${babelPath} src --quiet -d es --ignore "${ignoreGlobs}"`, {
     BABEL_ENV: 'es',
   });
-  exec(`${babelPath} src --quiet -d lib --ignore "${ignoreGlobs}"`, {
+  exec(`${babelPath} src -d lib --ignore "${ignoreGlobs}"`, {
```
Comment (abbeyhrt) [inline packages/react/scripts/build.js]: Out of curiosity, what's the reason for this change?

**Comment by abbeyhrt** [comment]
Hi! @samgoulden28! 👋 Thank you for your contribution! Do you think you could change the PR title to something like `feat(progress-indicator): add fitText prop and functionality to progress step` to more closely follow our commit conventions?

File: packages/react/scripts/build.js (hunk 1 above)
Comment (samgoulden28) [reply packages/react/scripts/build.js]: It was unintentional. Good spot, thanks!

**Comment by samgoulden28** [comment]
@abbeyhrt Thanks for your comments! Made your suggested change.

@laurenmrice, I added `interactive-04` as suggested,

What you have suggested for the spacing can be achieved but the only way I can think is some pretty ugly Javascript that will generate all elements, let the CSS/Sass calculate what the width will be for each and then go back over each of them and modify the styling on them adding the largest width.

However, I would not like to go down that route if I can avoid it so I will suggest some counter-offers:

1. We simply increase the width of the labels so that they can accommodate more text. They always will match and be equal although they will always be fixed.
2. Encourage users to introduce a width on the parent class. See the following screenshots:

width: 100%
image

width: 50%
image

We let flex-box do the rest. It will share out the remaining space between all of the flexing elements and you can decide when you use the component how much space to give it.

Let me know your comments and I would appreciate any other feedback on my suggestions. 

Many Thanks.

**Comment by laurenmrice** [comment]
@samgoulden28  thanks for your suggestions! the 50% width looks more accurate to me, I would be fine with showing that.

**Comment by samgoulden28** [comment]
So that would be a class that the user can introduce to the
ProgressIndicator element. I am happy to deliver a change to the storybook
that presents it as such. Is that ok?

**Comment by laurenmrice** [comment]
yes i think that is good @samgoulden28

**Comment by samgoulden28** [comment]
ALl updated @laurenmrice, Are you good to approve?

**Comment by tw15egan** [comment]
I might be a bit confused about what this PR is trying to achieve, but I'm still getting truncation while working on my MacBook with a smaller browser window...

Screen Shot 2020-02-27 at 1 02 24 PM

Is the goal to have it so that a user would specify something other than `width: 50%` on the parent container?  I think it may be more helpful to use the existing storybook example, and add a knob for the `fitText` prop? Although even then I'm seeing truncation if I add `fitText` to the existing example. 

Screen Shot 2020-02-27 at 1 07 58 PM

**Comment by samgoulden28** [comment]
@tw15egan The idea is to make it so that the labels can be bigger and display more text. I am trying to keep inline with request to make it so that all the labels are the same size so I decided to introduce flex items for the labels and have each item split the space equally. The idea of putting 50% on the parent in the storybook is to simulate what a user would do so that they could shorten the amount of space that is distributed between all the labels.

Perhaps fitText is not the correct propname now, and we should call it something like spaceEqually or something.

**Comment by tw15egan** [comment]
@samgoulden28 Gotcha! Yeah, I think the prop naming is what was confusing me. I think `spaceEqually` fits the description a bit better and may help someone who won't have this background with them 👍

**Comment by samgoulden28** [comment]
@tw15egan Take a look now buddy! I added a Knob like you suggested. See what you think. Cheers.

File: packages/react/src/components/ProgressIndicator/ProgressIndicator-story.js [hunk 2]
```diff
@@ -36,7 +47,9 @@ storiesOf('ProgressIndicator', module)
             <Tooltip
               direction="bottom"
               showIcon={false}
-              triggerClassName={`${prefix}--progress-label`}
+              triggerClassName={`${prefix}--progress-label ${spaceEquallyKnob() &&
+                !verticalKnob() &&
+                `${prefix}--progress-label--no-max-width`}`}
```
Comment (tw15egan) [inline packages/react/src/components/ProgressIndicator/ProgressIndicator-story.js]: Hmm, this is looking a bit gnarly, and it seems like every end-user would have to manually add this to any tooltip they render inside a tooltip to get `spaceEqually` to work properly, right?

I'm wondering if we can just modify the styles with a single, global style instead of a couple of different classes.

File: packages/components/src/components/progress-indicator/_progress-indicator.scss [hunk 3]
```diff
@@ -34,6 +34,11 @@
     }
   }
 
+  .#{$prefix}--progress-step--no-max-width {
```
Comment (tw15egan) [inline packages/components/src/components/progress-indicator/_progress-indicator.scss]: Based on the comment below, I wonder if we could add a global class like `bx--progress--max-width` and then just target the normal class names. This may help simplify the end-user experience/storybook example.

```scss
.#{$prefix}--progress--max-width .#{$prefix}--progress-step {
    min-width: 8rem;
    flex-grow: 1;
}
```

That way, in React, we would just have a global switch like

```jsx
const classes = classnames({
      [`${prefix}--progress`]: true,
      [`${prefix}--progress--vertical`]: vertical,
      [`${prefix}--progress--space-equal`]: !vertical && spaceEqually,
      [className]: className,
    });
```

File: packages/react/src/components/ProgressIndicator/ProgressIndicator.js [hunk 4]
```diff
@@ -117,13 +119,18 @@ export function ProgressStep({
           invalid={invalid}
           prefix={prefix}
         />
-        <ProgressStepLabel className={`${prefix}--progress-label`}>
+        <ProgressStepLabel
+          className={`${prefix}--progress-label ${spaceEqually &&
+            prefix + '--progress-label--no-max-width'}`}>
```
Comment (tw15egan) [inline packages/react/src/components/ProgressIndicator/ProgressIndicator.js]: If we use a global modifier, we can avoid having to attach classNames to each element that needs to be updated, and don't have to track the prop to each individual element

**Comment by samgoulden28** [comment]
An excellent suggestion @tw15egan !! I have made the changes and it works like a charm.

**Review by emyarod — COMMENTED** [review]
just a heads up, you will need to update Jest snapshots to clear the CI errors

File: packages/react/src/components/ProgressIndicator/ProgressIndicator-story.js [hunk 5]
```diff
@@ -16,14 +16,25 @@ import { settings } from 'carbon-components';
...
+
 storiesOf('ProgressIndicator', module)
   .addDecorator(withKnobs)
   .add(
     'Default',
     () => (
       <ProgressIndicator
-        vertical={boolean('Vertical', false)}
-        currentIndex={number('Current progress (currentIndex)', 1)}>
+        vertical={verticalKnob()}
```
Comment (tw15egan) [inline packages/react/src/components/ProgressIndicator/ProgressIndicator-story.js:35]: One tiny comment, is there any benefit to adding the knob this way? Otherwise, it may be nice to be consistent through our stories and just keep it as `vertical={boolean('Vertical', false)}`

File: packages/react/src/components/ProgressIndicator/ProgressIndicator-story.js [hunk 6]
```diff
@@ -16,14 +16,25 @@ import { settings } from 'carbon-components';
...
+        currentIndex={number('Current progress (currentIndex)', 1)}
+        spaceEqually={spaceEquallyKnob()}>
```
Comment (tw15egan) [inline packages/react/src/components/ProgressIndicator/ProgressIndicator-story.js:37]: Same as above, is there a benefit to this rather than doing `spaceEqually={boolean('Space Equally', false)}`?

File: packages/react/src/components/ProgressIndicator/ProgressIndicator.js [hunk 7]
```diff
@@ -285,10 +290,20 @@ export class ProgressIndicator extends Component {
...
+    /* eslint-disable no-unused-vars */
+    const {
+      className,
+      currentIndex,
+      spaceEqually,
+      vertical,
+      ...other
+    } = this.props;
+    /* eslint-enable no-unused-vars */
+    /* Only add spaceEqually if vertical is not true! */
```
Comment (tw15egan) [inline packages/react/src/components/ProgressIndicator/ProgressIndicator.js:302]: Can we move this comment down to `L306`?

File: packages/components/src/components/progress-indicator/_progress-indicator.scss [hunk 8]
```diff
@@ -70,6 +80,11 @@
     }
   }
 
+  .#{$prefix}--progress--space-equal .#{$prefix}--progress-label {
+    max-width: 100%;
```
Comment (ConradSchmidt) [inline packages/components/src/components/progress-indicator/_progress-indicator.scss:84]: Because I was looking into this feature as well; be aware that the hover state for the steps are a little bit off:
image

**Comment by joshblack** [comment]
Closing in favor of #5635
